v3/mqtt.go

```diff
@@ -465,91 +577,171 @@ func (conn *mqttConn) runPacketWriter(wg *sync.WaitGroup) {
+				// Case where the channel is closed
+				shouldLoop = false
+				break
+			}
+			if conn.verifyPacketType(pktSendData.pkt.Type()) != directPacketSendType {
+				resultCh <- MqttResponse{
+					Err: errors.New("Packet type must be queued"),
+				}
+			} else {
+				resultCh <- MqttResponse{Err: conn.writePacket(pktSendData.pkt)}
+			}
+		case pktSendData := <-conn.queuedSendPktCh:
+			if pktSendData.pkt == nil {
+				// Case where the channel is closed
+				shouldLoop = false
+				break
+			}
+			pktID := uint16(0)
+			if pktSendData.pkt.Type() == packet.PUBLISH {
+				pubPkt := pktSendData.pkt.(*packet.PublishPacket)
+				pktID = pubPkt.PacketID
+			}
+			// Get the packet ID, if any, for errors
+			resultCh := pktSendData.resultCh
+			if conn.verifyPacketType(pktSendData.pkt.Type()) != queuedPacketSendType {
+				resultCh <- MqttResponse{
+					Err: errors.New("Packet type must be sent directly"),
+				}
+			} else {
+				err := conn.writePacket(pktSendData.pkt)
+				if err != nil {
+					// If there was an error sending, we can notify the caller
+					// immediately.
+					resultCh <- MqttResponse{
+						PacketID: pktID,
+						Err:      err,
+					}
+				}
 			}
 		}
 	}
 }
 
+// Returns the channel to send the response, and the response data. If this
+// gets complicated, we can handle each type separately
+func (conn *mqttConn) createResponseForPacket(p packet.Packet) MqttResponse {
+	switch p.Type() {
+	case packet.PINGRESP:
+		// successful ping response is just nil errors
+		return MqttResponse{Err: nil}
+	case packet.PUBACK:
+		pubAck := p.(*packet.PubackPacket)
+		return MqttResponse{PacketID: pubAck.PacketID, Err: nil}
+	case packet.CONNACK:
+		connAck := p.(*packet.ConnackPacket)
+		var err error = nil
+		if connAck.ReturnCode != packet.ConnectionAccepted {
+			err = connAck.ReturnCode
+		}
+		return MqttResponse{Err: err}
+	case packet.SUBACK:
+		subAck := p.(*packet.SubackPacket)
+		resp := MqttResponse{
+			// We have do asynchronous SUBSCRIPTIONS, so our packets won't
+			// have packet ID's. But, if we ever do them, this is one place that
+			// we won't have to change our code.
+			PacketID: subAck.PacketID,
+		}
+		for i, code := range subAck.ReturnCodes {
+			if code == packet.QOSFailure {
+				err := errors.New("subscription rejected")
+				if i == 0 {
+					// If someone just checks Err of the response, at least
+					// they'll know that there was a failur
+					resp.Errs = append(resp.Errs, err)
+				}
+			}
+		}
+		return resp
+	default:
+		logError("Unhandled packet type for response: ", p.Type())
+		return MqttResponse{}
+	}
+}
+
 func (conn *mqttConn) runPacketReader(wg *sync.WaitGroup) {
 	defer func() {
 		logInfo("[MQTT] packet reader is exiting")
 		wg.Done()
-		// Close the rest of our channels?
 	}()
 
 	for {
+		// Set a deadline for reads to prevent blocking forever. We'll handle
+		// this error and continue looping, if appropriate
+		conn.conn.SetReadDeadline(time.Now().Add(time.Millisecond * 500))
```

Comment:
It better to define constant variable, I think.

Reply:
True. I was testing from the old code, and just copied the same value. I'll fix it.